### PR TITLE
updated findDescendantModels to add ability to filter the results

### PR DESCRIPTION
**Pull request by moloko**

### CODE REVIEW

File: src/core/js/models/adaptModel.js [hunk 1]
```diff
@@ -287,24 +289,42 @@ define([
...
+            } else {
+                searchChildren(this.getChildren().models);
+            }
 
-            return returnedDescendants;
+            if (!options) {
+                return returnedDescendants;
+            }
+
+            if (options.where) {
```
Comment (tomgreenfield) [inline src/core/js/models/adaptModel.js]: Please can we return early:
```js
var properties = options.where;

if (!properties) return;
```

File: src/core/js/models/adaptModel.js [hunk 2]
```diff
@@ -287,24 +289,42 @@ define([
...
+                return _.filter(returnedDescendants, function(descendant) {
```
Comment (tomgreenfield) [inline src/core/js/models/adaptModel.js]: Can use native `.filter()`:
```js
return returnedDescendants.filter(function(descendant) {
```

File: src/core/js/models/adaptModel.js [hunk 3]
```diff
@@ -287,24 +289,42 @@ define([
...
+                    for (var property in options.where) {
```
Comment (tomgreenfield) [inline src/core/js/models/adaptModel.js]: Are we particularly against using Underscore here? Should readability outweigh being performant?
```js
return _.findWhere(descendant, properties);
```

File: src/core/js/models/adaptModel.js (hunk 1 above)
Comment (moloko) [reply src/core/js/models/adaptModel.js]: @tomgreenfield the reason I didn't return early here is that there's an expectation that further `options` will be added in future - see [this suggestion](https://github.com/adaptlearning/adapt_framework/issues/2058#issuecomment-385449881) from @oliverfoster

File: src/core/js/models/adaptModel.js (hunk 3 above)
Comment (moloko) [reply src/core/js/models/adaptModel.js]: [I did already have a bit of a look at this](https://github.com/adaptlearning/adapt_framework/issues/2058#issuecomment-386673476) and felt that on balance `for...in` was the better option

**Pull request merged**